File: EyeTracker/CalibrationUtils.swift
//
//  CalibrationUtils.swift
//  EyeTracker
//
//  Created by Victor Schneuwly on 28.04.2024.
//

import ARKit

enum CalibrationState {
    case base
    case calibration(CGPoint, HeadPosition, PositionToScreen)
    case done
}

enum HeadPosition: String, CaseIterable {
    case middle = "Middle"
    case top = "Top"
    case down = "Down"
    case left = "Left"
    case right = "Right"

    func next() -> HeadPosition {
        let allCases = HeadPosition.allCases
        let currentIndex = allCases.firstIndex(of: self)!
        return allCases[(currentIndex + 1) % allCases.count]
    }

    func instruction() -> String {
        switch self {
        case .middle:
            return "Face the middle of the screen while looking at the target."
        case .top:
            return "Move your head up while looking at the target."
        case .down:
            return "Move your head down while looking at the target."
        case .left:
            return "Move your head to the left while looking at the target."
        case .right:
            return "Move your head to the right while looking at the target."
        }
    }
}

enum PositionToScreen: String, CaseIterable {
    case regular = "Regular"
    case armsExtended = "Arms Extended"
    case close = "Close"

    func next() -> PositionToScreen {
        let allCases = PositionToScreen.allCases
        let currentIndex = allCases.firstIndex(of: self)!
        return allCases[(currentIndex + 1) % allCases.count]
    }

    func instruction() -> String {
        switch self {
        case .regular:
            return "Hold your device at a regular distance from your face."
        case .armsExtended:
            return "Hold your device at arms extended distance from your face."
        case .close:
            return "Hold your device close to your face."
        }
    }
}

// MARK: - Calibration Data

struct CalibrationData {
    let username: String
    let deviceName: String
    let position: HeadPosition
    let distance: PositionToScreen
    let timestamp: Date
    let targetPoint: CGPoint
    let gazePoint: CGPoint
    let faceTransform: simd_float4x4
    let rightEyeTransform: simd_float4x4
    let leftEyeTransform: simd_float4x4
    let lookAtPoint: simd_float3

    static let csvHeader =
        "username, deviceName, position, distance, timestamp,"
            + "targetPointX, targetPointY, gazePointX, gazePointY,"
            + "faceTransform_0_x, faceTransform_0_y, faceTransform_0_z, faceTransform_0_w,"
            + "faceTransform_1_x, faceTransform_1_y, faceTransform_1_z, faceTransform_1_w,"
            + "faceTransform_2_x, faceTransform_2_y, faceTransform_2_z, faceTransform_2_w,"
            + "faceTransform_3_x, faceTransform_3_y, faceTransform_3_z, faceTransform_3_w,"
            + "rightEyeTransform_0_x, rightEyeTransform_0_y, rightEyeTransform_0_z, rightEyeTransform_0_w,"
            + "rightEyeTransform_1_x, rightEyeTransform_1_y, rightEyeTransform_1_z, rightEyeTransform_1_w,"
            + "rightEyeTransform_2_x, rightEyeTransform_2_y, rightEyeTransform_2_z, rightEyeTransform_2_w,"
            + "rightEyeTransform_3_x, rightEyeTransform_3_y, rightEyeTransform_3_z, rightEyeTransform_3_w,"
            + "leftEyeTransform_0_x, leftEyeTransform_0_y, leftEyeTransform_0_z, leftEyeTransform_0_w,"
            + "leftEyeTransform_1_x, leftEyeTransform_1_y, leftEyeTransform_1_z, leftEyeTransform_1_w,"
            + "leftEyeTransform_2_x, leftEyeTransform_2_y, leftEyeTransform_2_z, leftEyeTransform_2_w,"
            + "leftEyeTransform_3_x, leftEyeTransform_3_y, leftEyeTransform_3_z, leftEyeTransform_3_w"
            + "lookAtPointX, lookAtPointY, lookAtPointZ"

    func csvRepresentation() -> String {
        func extractValues(from transform: simd_float4x4) -> [Float] {
            return [
                transform.columns.0.x, transform.columns.0.y, transform.columns.0.z, transform.columns.0.w,
                transform.columns.1.x, transform.columns.1.y, transform.columns.1.z, transform.columns.1.w,
                transform.columns.2.x, transform.columns.2.y, transform.columns.2.z, transform.columns.2.w,
                transform.columns.3.x, transform.columns.3.y, transform.columns.3.z, transform.columns.3.w,
            ]
        }

        let faceTransformValues = extractValues(from: faceTransform)
        let rightEyeTransformValues = extractValues(from: rightEyeTransform)
        let leftEyeTransformValues = extractValues(from: leftEyeTransform)

        let values = [
            username, deviceName, position.rawValue, distance.rawValue, timestamp.timeIntervalSince1970,
            targetPoint.x, targetPoint.y, gazePoint.x, gazePoint.y,
        ] + faceTransformValues + rightEyeTransformValues + leftEyeTransformValues
            + [lookAtPoint.x, lookAtPoint.y, lookAtPoint.z]

        return values.map { "\($0)" }.joined(separator: ",")
    }
}

File: Notebooks/calibrators/calibration_utils.py
from dataclasses import dataclass
from enum import Enum
from typing import Tuple

import pandas as pd
import numpy as np


class PositionToScreen(Enum):
    REGULAR = "Regular"
    CLOSE = "Close"
    ARMS_EXTENDED = "Arms Extended"


class HeadPosition(Enum):
    MIDDLE = "Middle"
    TOP = "Top"
    DOWN = "Down"
    LEFT = "Left"
    RIGHT = "Right"


@dataclass
class CalibrationData:
    username: str
    device_name: str
    position: HeadPosition
    distance: PositionToScreen
    timestamp: str
    target_point: Tuple[float, float]
    gaze_point: Tuple[float, float]
    face_transform: np.ndarray  # 4x4 matrix
    right_eye_transform: np.ndarray  # 4x4 matrix
    left_eye_transform: np.ndarray  # 4x4 matrix
    euler_angles: Tuple[float, float, float]
    look_at_point: Tuple[float, float, float]


def import_from_csv(filename: str) -> [CalibrationData]:
    # Read the file and return a list of CalibrationData objects
    with open(filename, "r") as file:
        lines = file.readlines()
        # we can remove the first line as it contains the headers
        lines.pop(0)

        return [extract_calibration_data_from_line(line) for line in lines]


def extract_calibration_data_from_line(line: str) -> CalibrationData:
    # Extract the data from the line and return a CalibrationData object
    # we can split the line by comma
    data = line.split(",")

    return CalibrationData(
        username=data[0],
        device_name=data[1],
        position=HeadPosition(data[2]),
        distance=PositionToScreen(data[3]),
        timestamp=data[4],
        target_point=(float(data[5]), float(data[6])),
        gaze_point=(float(data[7]), float(data[8])),
        face_transform=np.array([float(x) for x in data[9:25]]).reshape((4, 4)),
        right_eye_transform=np.array([float(x) for x in data[25:41]]).reshape((4, 4)),
        left_eye_transform=np.array([float(x) for x in data[41:57]]).reshape((4, 4)),
        euler_angles=(float(data[57]), float(data[58]), float(data[59])),
        look_at_point=(float(data[60]), float(data[61]), float(data[62])),
    )


def import_from_dataframe(df) -> [CalibrationData]:
    # Extract the data from the DataFrame and return a list of CalibrationData objects
    return df.apply(extract_calibration_data_from_row, axis=1).tolist()


def extract_calibration_data_from_row(row) -> CalibrationData:
    return CalibrationData(
        username=row["username"],
        device_name=row["deviceName"],
        position=HeadPosition(row["position"]),
        distance=PositionToScreen(row["distance"]),
        timestamp=row["timestamp"],
        target_point=(float(row["targetPointX"]), float(row["targetPointY"])),
        gaze_point=(float(row["gazePointX"]), float(row["gazePointY"])),
        face_transform=np.array(
            row[
                [
                    f"faceTransform_{i}_{j}"
                    for i in range(4)
                    for j in ["x", "y", "z", "w"]
                ]
            ].tolist()
        ).reshape((4, 4)),
        right_eye_transform=np.array(
            row[
                [
                    f"rightEyeTransform_{i}_{j}"
                    for i in range(4)
                    for j in ["x", "y", "z", "w"]
                ]
            ].tolist()
        ).reshape((4, 4)),
        left_eye_transform=np.array(
            row[
                [
                    f"leftEyeTransform_{i}_{j}"
                    for i in range(4)
                    for j in ["x", "y", "z", "w"]
                ]
            ].tolist()
        ).reshape((4, 4)),
        euler_angles=(float(row["roll"]), float(row["pitch"]), float(row["yaw"])),
        look_at_point=(
            float(row["lookAtPointX"]),
            float(row["lookAtPointY"]),
            float(row["lookAtPointZ"]),
        ),
    )


def calib_data_to_dataframe(data):
    # turn a list of CalibrationData objects into a DataFrame
    return pd.DataFrame(
        [
            {
                "username": d.username,
                "deviceName": d.device_name,
                "position": d.position.value,
                "distance": d.distance.value,
                "timestamp": d.timestamp,
                "targetPointX": d.target_point[0],
                "targetPointY": d.target_point[1],
                "gazePointX": d.gaze_point[0],
                "gazePointY": d.gaze_point[1],
                **{
                    f"faceTransform_{i}_{j}": d.face_transform[i, ji]
                    for i in range(4)
                    for (ji, j) in enumerate(["x", "y", "z", "w"])
                },
                **{
                    f"rightEyeTransform_{i}_{j}": d.right_eye_transform[i, ji]
                    for i in range(4)
                    for (ji, j) in enumerate(["x", "y", "z", "w"])
                },
                **{
                    f"leftEyeTransform_{i}_{j}": d.left_eye_transform[i, ji]
                    for i in range(4)
                    for (ji, j) in enumerate(["x", "y", "z", "w"])
                },
                "roll": d.euler_angles[0],
                "pitch": d.euler_angles[1],
                "yaw": d.euler_angles[2],
                "lookAtPointX": d.look_at_point[0],
                "lookAtPointY": d.look_at_point[1],
                "lookAtPointZ": d.look_at_point[2],
            }
            for d in data
        ]
    )
